### DISCUSSION TITLE
How to 'merge' 2 duplicate objects and update all FK references?

### DISCUSSION THREAD

**Discussion by jsormaz** [sqlalchemy/sqlalchemy#11022] [Usage Questions]
Does there exist a mechanism to merge 2 objects?

I'm having difficulty searching because in sqlalchemy "merge" typically refers to session.merge. Maybe there is a better word to describe what I want to do.

for example, we have a 'Address' table with a couple entries
```python
a1 = Address(street='1234 Fake St.')
a2 = Address(street='1234 Fake St.')

c1 = Company(billing_address = a1, shipping_address = a2)

#some fictional 'merge' function (not to be confused with session.merge)
merge(a1,a2)
```

is there a dynamic way to "merge" a1 and a2 together, so that all fk references (in this case, Company.billing_address_id and Company.shipping_address_id) point to the same Address.id?

this is just a simplified example, in reality of course there are way more tables and relationships, that make it a little less reliable to manually walk through and set.

**Comment by zzzeek** [comment]
well if one of them were present in the `Session`, doing a `Session.merge()` with the other one would in fact overwrite all the attributes with what's incoming.

otherwise just write a function to do what you need

**Reply by jsormaz** [reply]
Sorry, i'm not so much trying to update the attributes of one of the objects.
I want to delete 1 of the objects, and have any related objects that referenced the deleted object be updated to reference the retained object,

so more like a 'merge identity' than 'merge data'

**Reply by zzzeek** [reply]
if these are one to manys or many-to-manys you would move them over something like this

```
my_new_object.list_of_things.extend(my_old_object.list_of_things)
```

**Reply by CaselIT** [reply]
I would do the opposite:
- look if the address exists
- use the existing address from the session
- if not create a new one and use that

so something like this:
```py
def get_address(street, session):
  stmt = sa.select(Address).where(Address.street=street)
  address = session.execute(stmt).one_or_none()
  if address is None:
    address = Address(street=street)
    session.add(address)
  return address
```

**Reply by jsormaz** [reply]
@CaselIT for new entries I agree that is the way, and I have some logic in place to do that for inserts.

In my case, the database is already a rotten mess of duplicated information though (migrated from 2 old systems over some time period)

I'm trying to provide a way for the user to select some addresses and say "squash these n addresses into 1 db row"

I was hoping there would be a way to basically "find all columns which have a foreignkey constraint to a1.id and replace with a2.id" without having to manually go through the relationships.
I'm sure I can still do something semi-automated to iterate through relationships, I just worry there is a chance of missing one (for example a one sided relationship (without backref/back_populates) that Address model knows nothing about) , and then it gets orphaned when I delete a2.

I appreciate the comments, this is helpful.

**Reply by CaselIT** [reply]
> I was hoping there would be a way to basically "find all columns which have a foreignkey constraint to a1.id and replace with a2.id" without having to manually go through the relationships.

no, I don't think there is anything automated, sorry.
